python/mxnet/gluon/trainer.py

```diff
@@ -187,13 +240,9 @@ def step(self, batch_size, ignore_stale_grad=False):
                             "warning and skip updating of Parameters with stale gradient" \
                             %(param.name, str(data.context)))
 
-            if self._kvstore:
-                self._kvstore.push(i, param.list_grad(), priority=-i)
-                if self._update_on_kvstore:
-                    self._kvstore.pull(i, param.list_data(), priority=-i)
-                    continue
-                else:
-                    self._kvstore.pull(i, param.list_grad(), priority=-i)
+            if self._kvstore and self._update_on_kvstore:
```

Comment:
A previous push is not guaranteed. I don't think it makes sense to allow users call allreduce and update separately with update_on_kvstore

Reply:
This is now handled in `update()`.

Reply:
"don't think it makes sense to allow users call allreduce and update separately with update_on_kvstore"
+1